src/storage/scripting.cc

```diff
@@ -1470,8 +1488,73 @@ Status CreateFunction(Server *srv, const std::string &body, std::string *sha, lu
   }
   lua_setglobal(lua, funcname);
 
+  // Cache the flags of the current script
+  uint64_t script_flags = 0;
+  if (auto pos = body.find('\n'); pos != std::string::npos) {
+    auto first_line = body.substr(0, pos);
+    if (util::HasPrefix(first_line, "#!lua")) {
+      uint64_t shebang_flags = 0;
+      if (auto s = ExtractFlagsFromShebang(first_line, shebang_flags); !s.IsOK()) {
+        lua_pop(lua, 1); /* remove the error handler from the stack. */
+        return s;
+      }
+      script_flags |= shebang_flags;
+    } else {
+      // scripts without #! can run commands that access keys belonging to different cluster hash slots,
+      // but ones with #! inherit the default flags, so they cannot.
+      script_flags |= ScriptFlags::kScriptAllowCrossSlotKeys;
+    }
+  }
+  srv->CacheScriptFlags(*sha, script_flags);
+
   // would store lua function into propagate column family and propagate those scripts to slaves
   return need_to_store ? srv->ScriptSet(*sha, body) : Status::OK();
 }
 
+[[nodiscard]] Status ExtractLibNameFromShebang(const std::string &shebang, std::string &libname) {
+  static constexpr const char *shebang_prefix = "#!lua";
+  static constexpr const char *shebang_libname_prefix = "name=";
+
+  if (!util::HasPrefix(shebang, shebang_prefix)) {
+    return {Status::NotOK, "Expect shebang prefix \"#!lua\" at the beginning of the first line"};
+  }
+
+  if (auto pos = shebang.find(shebang_libname_prefix, strlen(shebang_prefix)); pos != std::string::npos) {
```

Comment:
here we need to consider more to prevent invalid input, e.g. `#!luaname=..`, `#!lua xxxname=..`, or valid input with a key that's currently undefined, e.g. `#!lua name=a key1=b`.